specification/src/instructions.md

```diff
@@ -1,182 +1,811 @@
+secret – execution _is_ deterministic: the divined values were supplied as secret input and are read
+from there.
+
+### Hashing and Sponge Instructions
+
+The instructions [`sponge_init`](#sponge_init), [`sponge_absorb`](#sponge_absorb),
+[`sponge_absorb_mem`](#sponge_absorb_mem), and [`sponge_squeeze`](#sponge_squeeze) are the interface
+for using the [Tip5](https://eprint.iacr.org/2023/107.pdf) permutation in a
+[Sponge](https://keccak.team/sponge_duplex.html) construction.
+The capacity is never accessible to the program that's being executed by Triton VM.
+
+At any given time, at most one Sponge state exists.
+In particular, Triton VM does not start with an initialized Sponge state.
+If a program uses the Sponge state, the first Sponge instruction must be `sponge_init`;
+otherwise, Triton VM will crash.
 
-## Regarding Opcodes
+Only instruction `sponge_init` resets the state of the Sponge, and only the three Sponge
+instructions influence the Sponge's state.
+Notably, executing instruction [`hash`](#hash) does not modify the Sponge's state.
 
-An instruction's _[operation code](https://en.wikipedia.org/wiki/Opcode)_, or _opcode_, is the machine word uniquely identifying the instruction.
-For reasons of efficient [arithmetization](arithmetization.md), certain properties of the instruction are encoded in the opcode.
+When using the Sponge instructions, it is the programmer's responsibility to take care of proper
+input padding:
+Triton VM cannot know the number of elements that will be absorbed.
+For more information, see section 2.5 of the [Tip5 paper](https://eprint.iacr.org/2023/107.pdf).
+
+### Regarding Opcodes
+
+An instruction's _[operation code](https://en.wikipedia.org/wiki/Opcode)_, or _opcode_, is the
+machine word uniquely identifying the instruction.
+For reasons of efficient [arithmetization](arithmetization.md), certain properties of the
+instruction are encoded in the opcode.
 Concretely, interpreting the field element in standard representation:
 - for all double-word instructions, the least significant bit is 1.
 - for all instructions shrinking the operational stack, the second-to-least significant bit is 1.
-- for all [u32 instructions ](u32-table.md), the third-to-least significant bit is 1.
+- for all [u32 instructions](#bitwise-arithmetic-on-stack), the third-to-least significant bit is 1.
+
+The first property is used by instruction [skiz](#skiz) (see also its
+[arithmetization](instruction-specific-transition-constraints.md#instruction-skiz)).
+The second property helps with proving consistency of the
+[Op Stack](data-structures.md#operational-stack).
+The third property allows efficient arithmetization of the running product for the
+Permutation Argument between [Processor Table](processor-table.md) and [U32 Table](u32-table.md).
+
+## Summary
+
+The following table is a summary of all instructions.
+For more details, read on below.
```

Comment:
Summary is really helpful! Would you also mention before the table all the sectioning used below? (As we don't know if we will ever have a ToC here.) I feel the minimal comma separated naming of each section in a line would be enough, though there are nicer ways too. X)

Reply:
Good idea. Done. :+1: